### DISCUSSION TITLE
The best way to store the data?

### DISCUSSION THREAD

**Discussion by dmarklund** [ueberdosis/tiptap#148] [General Questions]
Just like the title states, which is the best data type to store, the JSON or generated HTML?
It's always best practice to store the users raw input, but storing both seems a bit overkill.

Thanks for the great work on the editor!

**Comment by cjaoude** [comment]
You store both. At some point you have to display the html version. If it is not persisted, the app has to parse the JSON version each and every time. Think of the html version as the _read-only_ version. The JSON version is what you load into the editor or use for transformations.

JSON and HTML are text. You can store oodles of it in the db until the cows come home.

**Comment by philippkuehn** [comment]
It depends. For simple projects HTML should be fine. In our case at scrumpy we store JSON because we want to format the content in different ways, e.g. for emails (we built a PHP parser for it).

**Comment by dmarklund** [comment]
@philippkuehn Yeah,  it really does depend on the use case. Storing the plain JSON obviously gives you more flexibility to generate the HTML afterwards and displaying it differently. Will you release the parser too? 👏

**Comment by hanspagel** [comment]
I'll put it on GitHub!

**Comment by hanspagel** [comment]
It's still work in progress, but it's public now:
https://github.com/scrumpy/html-to-prosemirror

**Comment by dmarklund** [comment]
@hanspagel Awesome, thanks a bunch!

**Comment by hanspagel** [comment]
I'm working on publishing prosemirror-to-html (the other way around) too.

**Comment by dmarklund** [comment]
Sounds good. It could be nice to be able to convert the data all ways as you're mentioning. Something specific that you need help with?

**Comment by hanspagel** [comment]
Not really. We need to build those converters to migrate our WYSIWYG data to prosemirror/titap anyway.

Just let me know if something is missing/broken if you're using it.

**Comment by kinthaiofficial** [comment]
The best way to store Tiptap editor data depends on your use case, but the key decision is: **JSON (ProseMirror document) vs HTML**.

**JSON (recommended for most cases):**
```typescript
// Save
const docJson = editor.getJSON();
await saveToDatabase(docJson);

// Load
editor.commands.setContent(docJson);
```

**Why JSON is better:**
- Structure-preserving: marks, nodes, attributes are all represented accurately
- Renderable without the editor: use `generateHTML(docJson, extensions)` server-side
- Diffable: JSON diffs are meaningful; HTML diffs are brittle
- Collaborative editing: Yjs/Hocuspocus use JSON document structure, not HTML

**HTML storage (when you need it):**
```typescript
const html = editor.getHTML();
// Useful for: email templates, legacy system integration, static rendering
```

**The hybrid approach (for full-text search):**
Store JSON as the source of truth, store extracted plain text for search indexing:
```typescript
const json = editor.getJSON();
const text = editor.getText();  // For full-text search index

await db.documents.update({
  content: json,
  searchIndex: text,
  updatedAt: new Date(),
});
```

**Collaborative editing + persistence:**
If using Hocuspocus, the server handles persistence as a Yjs doc (binary format, more efficient for real-time collaboration). Export to JSON/HTML for static rendering or search.

**For AI-generated content:**
LLMs can generate Tiptap-compatible JSON if you include the schema in the prompt. This lets AI write structured documents (with bold, lists, code blocks) rather than just plain text. Alternatively, generate Markdown and convert via `prosemirror-markdown`.

**Comment by kinthaiofficial** [comment]
The storage format decision in Tiptap is one of the most important architectural choices for a content-heavy application. Here's the breakdown:

**Option 1: JSON (Tiptap's native format)** — Recommended for most cases:
```typescript
// Save
const json = editor.getJSON();
await db.content.save({ body: json });  // Store as JSONB in Postgres

// Load
editor.commands.setContent(savedJson);

// Query specific content (with Postgres JSONB):
const posts = await db.query(
  `SELECT * FROM posts WHERE body @> '{"type": "doc"}'`
);
```

Pros: Preserves all Tiptap metadata, easy to re-render, queryable with JSONB, extensible.
Cons: Slightly larger storage, not human-readable.

**Option 2: HTML** — Good for interoperability:
```typescript
const html = editor.getHTML();
// Store as TEXT in DB

editor.commands.setContent('<p>Hello world</p>');
```

Pros: Portable, renders in any context, search-indexable.
Cons: Tiptap-specific extensions may produce non-standard HTML, some metadata lost.

**Option 3: Markdown** — Only if users expect markdown workflow:
```typescript
// Requires @tiptap/extension-markdown
import Markdown from 'tiptap-markdown';
const md = editor.storage.markdown.getMarkdown();
```

Pros: Human-readable, version control friendly, search-friendly.
Cons: Lossy (complex formatting lost), not all extensions have markdown equivalents.

**Best practice** for collaborative apps:
- Store JSON as primary format
- Generate HTML on save for search indexing + read-only views
- Consider Yjs (via `@tiptap/extension-collaboration`) if you need real-time collaboration — adds a CRDT layer

```typescript
// The database schema that works well:
// content_body: JSONB        (full fidelity, for editing)
// content_text: TEXT         (plain text extracted, for search)
// content_html: TEXT         (rendered HTML, for read views)
```
